draft-ietf-gnap-core-protocol.md

```diff
@@ -307,6 +307,30 @@ A global assumption made by GNAP is that authorization requests are security and
 
+## Protocol Flow {#protocol}
+
+GNAP is fundamentally designed to allow delegated access to APIs and other information, such as subject information, using a multi-stage, stateful process. This process allowing different parties to provide information into the system to alter and augment the state of the delegated access and its artifacts.
+
+The underlying requested grant moves through several states as different actions take place during the protocol:
+
+~~~ aasvg
+{::include diagram/states.md}
+~~~
+
+*Processing*
+: When a [request for access](#request) is received by the AS, a new grant request is in _processing_ by the system. At this stage, the AS processes the state of the request to determine whether interaction with the end user or RO is required for approval of the request. The grant request has to exit this state before a response can be returned to the client instance. If approval is required, the request moves to the _pending_ state and the AS returns a [continue response](#response-continue) along with any appropriate [interaction responses](#response-interact). If no such approval is required, such as when the client instance is acting on its own behalf or the AS can determine that access has been fulfilled, the request moves to the _approved_ state where [access tokens for API access](#response-token) and [subject information](#response-subject) can be issued to the client instance.
+
+*Pending*
+: When a request needs to be approved by a RO, or interaction with the end user is required, the grant request enters a state of _pending_. In this state, no access tokens can be granted and no subject information can be released to the client instance. While a grant request is in this state, the AS seeks to gather the required [consent and authorization](#authorization) for the requested access. A grant request in this state is always associated with a _continuation access token_ bound to the client instance's key. If no [interaction finish method](#request-interact-finish) is associated with this request, the client instance can send a [polling continue request](#continue-poll) to the AS. This returns a [continue response](#response-continue) while the grant request remains in this state, allowing the client instance to continue to check the state of the pending grant request. If an [interaction finish method](#request-interact-finish) is specified in the grant request, the client instance can [provide this](#continue-after-interaction) to the AS to move this request to the _processing_ state to be re-evaluated by the AS. When new information is available in the context of the grant request, the AS moves this request to the _processing_ state to be re-evaluated. This information can be supplied by the client in the form of a [post-interaction continuation request](#continue-after-interaction) when an [interaction finish method](#request-interact-finish) is associated with the grant request.
+
+*Approved*
```

Comment:
"Approved" is not a terminal state. Once something is approved it can be modified and later revoked.

"Rejected/denied" is an action, not a state, that puts the state back into "processing", where the AS determines what to do next. The AS could decide there's nothing else to do (the end user's decision is final) and go to "revoked", or it could move to "pending" if there's more that could happen to cause it to get approved, such as claims sent from the client software itself, or a separate interaction.

Reply:
@yaronf I've added more text that hopefully makes this distinction more clear.

Reply:
@jricher I was referring to the *AS* making a decision to reject the request.

And now I'm even more confused. If both the AS and the RO are happy, what is the terminal state that denotes that the request was successful? It cannot be "revoked", can it.